Added documentation for account notes (#474)
MewsSystems/gitbook-connector-api: operations/accountnotes.md

            "IsActive": true,
            "CreatorProfile": {
                "Discriminator": "Enterprise",
                "EnterpriseProfile": {
                    "ProfileId": "52d19c34-b0aa-4635-905d-1326fa8b8e13"
                }
            },
            "UpdaterProfile": {
                "Discriminator": "Integration",
                "EnterpriseProfile": null
            }
        }
    ],
    "Cursor": "da34b396-41f7-47f6-8847-aecf00a3f19e"
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `AccountNotes` | array of [Account note](#account-note) | required | The set of requested account notes. |
| `Cursor` | string | optional | Unique identifier of the item one newer in time order than the items to be returned. If Cursor is not specified, i.e. null, then the latest or most recent items will be returned. |

#### Account note

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Id` | string | required | Unique identifier of the account note. |
| `AccountId` | string | required | Unique identifier of the account. |
| `Content` | required | required | The content of the account note. |

> | `AccountType` | string [Account type](#account-type)| required | Specifying types of accounts provided ([Customer](../operations/customers.md#customer), [Company](../operations/companies.md#company)). |

Reviewer: What about `string [Account type](#account-type)` in the second column? 🤔

> | `AccountType` | string [Account type](#account-type)| required | Specifying types of accounts provided ([Customer](../operations/customers.md#customer), [Company](../operations/companies.md#company)). |

Reviewer 2: The link is there but it doesn't have anything to lead to, or am I missing something? It is not rendered as a link in a preview too.

> | `AccountType` | string [Account type](#account-type)| required | Specifying types of accounts provided ([Customer](../operations/customers.md#customer), [Company](../operations/companies.md#company)). |

Reviewer: Ah I got you now 👍🏻 Yeah it's missing the document name. I'm updating it as `accounts.md#account-type`

> | `AccountType` | string [Account type](#account-type)| required | Specifying types of accounts provided ([Customer](../operations/customers.md#customer), [Company](../operations/companies.md#company)). |

Reviewer: [Done](https://github.com/MewsSystems/gitbook-connector-api/pull/474/commits/4022a1024e8d383fd2cbe4a2fed57da7d3e5d6f6) 👍🏻

                "Value": "The AC in the room doesn't work anymore"
            },
            "Classifications": {
                "Maintenance": {
                    "Value": true
                },
                "Housekeeping": {
                    "Value": false
                }
            }
        }
    ]
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `ClientToken` | string | required | Token identifying the client application. |
| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |
| `AccountNoteUpdates` | array of [Account note update parameters](#account-note-update-parameters) | required, max 1000 items | Account notes to be updated. |

#### Account note update parameters

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `AccountNoteId` | string | required | Unique identifier of the account note. |
| `Content` | [String update value](_objects.md#string-update-value) | optional, max length 1000 characters | Content of the account note \(or `null` if the content should not be updated\). |

> | `Classifications` | [Customer note classifications update parameters](#customer-note-classifications-update-parameters) or [Company note classifications update parameters](#company-note-classifications-update-parameters) | required | Classification of the account note |

Reviewer 2: I missed this during the contract review for /update endpoint, but I believe I've mentioned it on /add review - it is not good that the deserialization depends on some implied value (here it is the type of account). Either it should have the same contract with different validation or an explicitly specified discriminator.

Reviewer: I think that's been updated when we pass to the update objects with bool values for classifications. Not sure if it makes sense to do it like that now 🤔 We would be making all customer-related-classifications available to companies in the request. Is that something we really want?
We will be evaluating the account type and classification type if it helps 😅)

Reviewer 2: The principle would be the same as in /add case - have a common list of classification values for both types of account in the contract, but restrict them using validation and also specify that in documentation. I'm not saying that it needs to be changed, but a possible way would be to have AccountNoteUpdateParameters like this:
```
{
     ...
     IEnumerable<EnumUpdateValue<AccountNoteClassification>> Classifications { get; set; }
}
```

where EnumUpdateValue<T> would be something along the lines of
```
{
      T Value { get; set; }
      bool IsSet { get; set; }
}
```

Reviewer: This actually would be nice for update scenario as well. But at some point there's been an agreement for list of enum strings for `/add`, and boolean values for `/update`.
So for `/add` you say no need to change it. What's the decision for `/update`? Should we update it with one classification object?

Reviewer 2: I don't think it's worth redoing, just leave it how it is.

# Account notes

## Get all account notes

> Returns all account notes of an account (in the given activity state), optionally filtered by account identifiers by specific account note identifiers or other filter parameters.

Reviewer 3: ```suggestion
Returns all account notes of an account, optionally filtered by activity state, account identifiers, specific account note identifiers or other filter parameters.
```

    "Client": "Sample Client 1.0.0",
    "ChainIds": [
        "1df21f06-0cfc-4960-9c58-a3bf1261663e",
        "5fcd1933-22f2-40b9-84da-7db04cbecec2"
    ],
    "AccountNoteIds": [
        "3ed9e2f3-4bba-4df6-8d41-ab1b009b6425",
        "8a98965a-7c03-48a1-a28c-ab1b009b53c8"
    ],
    "AccountIds": [
        "a6738390-c241-45b7-8e46-14f47207abe5",
        "435d4d5f-d14f-48dc-a47e-0481fc28ead0"
    ],
    "UpdatedUtc": {
        "StartUtc": "2022-10-10T00:00:00Z",
        "EndUtc": "2022-10-17T00:00:00Z"
    },
    "ActivityStates": [
        "Active"
    ],
    "Limitation":{
        "Cursor": "e7f26210-10e7-462e-9da8-ae8300be8ab7",
        "Count": 100
    }
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `ClientToken` | string | required | Token identifying the client application. |
| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |
| `ChainIds` | array of string | optional, max 1000 items | Unique identifiers of the chain. If not specified, the operation returns data for all chains within scope of the Access Token. |

> | `AccountNoteIds` | array of string | required, max 1000 items | Unique identifiers of [Account note](#account-note). |

Reviewer 3: Is `AccountNoteIds` **required**, as it states here? This is not what it says in the description for the operation.

            "CreatorProfile": {
                "Discriminator": "Enterprise",
                "EnterpriseProfile": {
                    "ProfileId": "52d19c34-b0aa-4635-905d-1326fa8b8e13"
                }
            },
            "UpdaterProfile": {
                "Discriminator": "Integration",
                "EnterpriseProfile": null
            }
        },
        {
            "Id": "da34b396-41f7-47f6-8847-aecf00a3f19e",
            "AccountId": "5fcd1933-22f2-40b9-84da-7db04cbecec2",
            "Content": "Lactose intolerant",
            "AccountType": "Customer",
            "Classifications": ["FoodAndBeverage"],
            "IsActive": true,
            "CreatorProfile": {
                "Discriminator": "Enterprise",
                "EnterpriseProfile": {
                    "ProfileId": "52d19c34-b0aa-4635-905d-1326fa8b8e13"
                }
            },
            "UpdaterProfile": {
                "Discriminator": "Integration",
                "EnterpriseProfile": null
            }
        }
    ],
    "Cursor": "da34b396-41f7-47f6-8847-aecf00a3f19e"
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `AccountNotes` | array of [Account note](#account-note) | required | The account note of the account within the chain. |

Reviewer 3: ```suggestion
| `AccountNotes` | array of [Account note](#account-note) | required | The set of requested account notes. |
```

| `Content` | required | required | The content of the account note. |
| `AccountType` | string [Account type](#account-type)| required | Specifying types of accounts provided ([Customer](../operations/customers.md#customer), [Company](../operations/companies.md#company)). |
| `Classifications` | array of string | required | Specifying the classifactions of the note based on account type ([Customer note classification](#customer-note-classifications), [Company note classifications](#company-note-classifications)). |
| `IsActive` | boolean | required | Whether the account note is till active. |
| `CreatorProfile` | [Profile data](#profile-data) | required | The profile data of the user who created the account note. |
| `UpdaterProfile` | [Profile data](#profile-data) | required | The profile data of the user who updated the account note. |

#### Customer note classifications

* `General`
* `FoodAndBeverage`
* `FrontOffice`
* `Reservations`
* `Housekeeping`
* `Maintenance`
* `PreviousStay`
* `FamilyRelations`
* `Gifts`
* `Accounting`
* `Complaints`
* `Other`

#### Company note classifications

* `General`

## Add account notes

> > This feature is being actively developed, features and behavior of this operation may change at short notice. Currently we support only one classification per account note.

Reviewer 3: ```suggestion
> ### Restricted!
> This feature is being actively developed, features and behavior of this operation may change at short notice. Currently we support only one classification per account note.
```

Reviewer: About this, I think we no longer need to add the message about "being actively developed" 🤔 
As far as I know, we don't have any upcoming change. I think we can get rid of that part completely. @jonas-paul that works for you too?

| `Content` | string | required, max length 1000 characters | Content of the account note. |
| `Classifications` | array of string | required, max length 1 | Classification of the account note based on account type ([Customer note classification](#customer-note-classifications), [Company note classifications](#company-note-classifications)). |

### Response

```javascript
{
    "AccountNotes": [
        {
            "Id": "a58ff7cb-77e3-495a-bd61-aecf00a3f19d",
            "AccountId": "1df21f06-0cfc-4960-9c58-a3bf1261663e",
            "Content": "Brother of the CEO",
            "AccountType": "Customer",
            "Classifications": ["FamilyRelations"],
            "IsActive": true,
            "CreatorProfile": {
                "Discriminator": "Enterprise",
                "EnterpriseProfile": {
                    "ProfileId": "52d19c34-b0aa-4635-905d-1326fa8b8e13"
                }
            },
            "UpdaterProfile": {
                "Discriminator": "Integration",
                "EnterpriseProfile": null
            }

>         },

Reviewer 3: ```suggestion
        }
```

    ],
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `AccountNotes` | array of [Account notes](#account-note) | required | Added account notes. |

## Update account notes

> > This feature is being actively developed, features and behavior of this operation may change at short notice. Currently we support only one classification per account note. When changing the classification of a note, the current classification must be removed.

Reviewer 3: ```suggestion
> ### Restricted!
> This feature is being actively developed, features and behavior of this operation may change at short notice. Currently we support only one classification per account note. When changing the classification of a note, the current classification must be removed.
```

Updates information about the specified account notes.

### Request

`[PlatformAddress]/api/connector/v1/accountNotes/update`

```javascript
{
    "ClientToken": "E0D439EE522F44368DC78E1BFB03710C-D24FB11DBE31D4621C4817E028D9E1D",
    "AccessToken": "C66EF7B239D24632943D115EDE9CB810-EA00F8FD8294692C940F6B5A8F9453D",
    "Client": "Sample Client 1.0.0",
    "AccountNoteUpdates": [
        {
            "AccountNoteId": "a58ff7cb-77e3-495a-bd61-aecf00a3f19d",
            "Content": {
                "Value": "The AC in the room doesn't work anymore"
            },
            "Classifications": {
                "Maintenance": {
                    "Value": true
                },
                "Housekeeping": {
                    "Value": false
                }

>             },

Reviewer 3: ```suggestion
            }
```

        }
    ]
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `ClientToken` | string | required | Token identifying the client application. |
| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |
| `AccountNoteUpdates` | array of [Account note update parameters](#account-note-update-parameters) | required, max 1000 items | Account notes to be updated. |

#### Account note update parameters

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `AccountNoteId` | string | required | Unique identifier of the account note. |

> | `Content` | [String update value](_objects.md#string-update-value) | optional, max length 1000 characters | Content of the account note \(or `null` if the name should not be updated\). |

Reviewer 3: Should that be "\(or `null` if the _content_ should not be updated\)?"

Reviewer: It should, it must be a copy-paste mistake. I'll fix it :)

> | `Classifications` | [Customer note classifications update parameters](#customer-note-classifications-update-parameters), [[Company note classifications update parameters](#company-note-classifications-update-parameters)] | required | Classification of the account note |

Reviewer 3: ```suggestion
| `Classifications` | [Customer note classifications update parameters](#customer-note-classifications-update-parameters) or [Company note classifications update parameters](#company-note-classifications-update-parameters) | required | Classification of the account note |
```

| :-- | :-- | :-- | :-- |
| `General` | [Bool update value](_objects.md#bool-update-value) | optional | Boolean value defining the general classification for the account note. \(or `null` if the value should not be updated\). |

### Response

```javascript
{
    "AccountNotes": [
        {
            "Id": "a58ff7cb-77e3-495a-bd61-aecf00a3f19d",
            "AccountId": "1df21f06-0cfc-4960-9c58-a3bf1261663e",
            "Content": "The AC in the room doesn't work anymore",
            "AccountType": "Customer",
            "Classifications": ["Maintenance"],
            "IsActive": true,
            "CreatorProfile": {
                "Discriminator": "Enterprise",
                "EnterpriseProfile": {
                    "ProfileId": "52d19c34-b0aa-4635-905d-1326fa8b8e13"
                }
            },
            "UpdaterProfile": {
                "Discriminator": "Integration",
                "EnterpriseProfile": null
            }
        },
    ],
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `AccountNotes` | array of [Account note](#account-note) | required | Updated account notes. |

## Delete account notes

> Deletes account notes. 

Reviewer 3: ```suggestion
Deletes specified account notes. 
```

### Request

`[PlatformAddress]/api/connector/v1/accountNotes/delete`

```javascript
{
    "ClientToken": "E0D439EE522F44368DC78E1BFB03710C-D24FB11DBE31D4621C4817E028D9E1D",
    "AccessToken": "C66EF7B239D24632943D115EDE9CB810-EA00F8FD8294692C940F6B5A8F9453D",
    "Client": "Sample Client 1.0.0",
    "AccountNoteIds": [
        "a58ff7cb-77e3-495a-bd61-aecf00a3f19d",
		"da34b396-41f7-47f6-8847-aecf00a3f19e"
    ]
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `ClientToken` | string | required | Token identifying the client application. |
| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |

> | `AccountNoteIds` | array of string | required, max 1000 items | Unique identifier of the account notes to be deleted. |

Reviewer 3: ```suggestion
| `AccountNoteIds` | array of string | required, max 1000 items | Unique identifiers of the account notes to be deleted. |
```
